geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/process/RouteSearchProcess.scala

```diff
@@ -0,0 +1,349 @@
+    } else if (headingField != null && sft.indexOf(headingField) == -1) {
+      throw new IllegalArgumentException(s"Heading field '$headingField' does not exist in input feature collection")
+    }
+
+    // extract the route geometries
+    val routeGeoms = {
+      val sft = routes.getSchema
+      val name = Option(routeGeomField).getOrElse(sft.getGeometryDescriptor.getLocalName)
+      val index = sft.indexOf(name)
+      if (index == -1) {
+        throw new IllegalArgumentException(s"Geometry field '$name' does not exist in route feature collection")
+      }
+      if (!classOf[LineString].isAssignableFrom(sft.getDescriptor(index).getType.getBinding)) {
+        throw new IllegalArgumentException(s"Route geometry field '$name' must be a LineString")
+      }
+      SelfClosingIterator(routes).map(_.getAttribute(index).asInstanceOf[LineString]).toSeq
+    }
+
+    val visitor = new RouteVisitor(routeGeoms, bufferSize, headingThreshold, bi, geomAttribute, isPoints, Option(headingField))
+
+    features.accepts(visitor, new NullProgressListener)
+    val results = visitor.getResult.results
+
+    configureOutput(features.getSchema, results, bins, binGeom, binDtg, binTrackId, binLabel)
+
+    results
+  }
+}
+
+class RouteVisitor(routes: Seq[LineString],
+                   routeBuffer: Double,
+                   threshold: Double,
+                   bidirectional: Boolean,
+                   geomAttribute: String,
+                   isPoints: Boolean,
+                   headingAttribute: Option[String]) extends FeatureCalc with FeatureAttributeVisitor with LazyLogging {
+
+  private var resultCalc: RouteResult = _
+
+  private val routeFilter =
+    orFilters(routes.map(ls => ff.dwithin(ff.property(geomAttribute), ff.literal(ls), routeBuffer, "meters")))
+
+  // for manual check, rewrite the filter to handle meters
+  // geotools doesn't handle this, we normally do it in our query planner
+  private lazy val manualRouteFilter = routeFilter match {
+    case f: DWithin => new QueryPlanFilterVisitor(null).visit(f, null).asInstanceOf[Filter]
+    case f: Or      => new QueryPlanFilterVisitor(null).visit(f, null).asInstanceOf[Filter]
+  }
+
+  // for collecting results manually
+  private var manualCollection: ListFeatureCollection = _
+
+  override def getResult: RouteResult = resultCalc
+
+  // manually called for non-accumulo feature collections
+  override def visit(feature: Feature): Unit = {
+    val sf = feature.asInstanceOf[SimpleFeature]
+
+    if (manualCollection == null) {
+      manualCollection = new ListFeatureCollection(sf.getFeatureType)
+      resultCalc = RouteResult(matchRoutes(manualCollection))
+    }
+
+    if (manualRouteFilter.evaluate(sf)) {
+      manualCollection.add(sf)
+    }
+  }
+
+  // allows us to accept visitors from retyping feature collections
+  override def getExpressions: java.util.List[Expression] = {
```

Comment:
not sure actually - this lets it know which columns you want. if you don't implement this interface, the RetypingFeatureCollection will never delegate. If you do implement it, then it will compare the columns you want with the underlying columns and decide if it can delegate or not. Possibly implementing it but returning an empty list would always let it delegate...